torch/csrc/cuda/green_context.h

```diff
@@ -0,0 +1,191 @@
+      : green_ctx_(other.green_ctx_),
+        context_(other.context_),
+        parent_stream_(other.parent_stream_) {
+    other.green_ctx_ = nullptr;
+    other.context_ = nullptr;
+    other.parent_stream_ = NULL;
+  }
+
+  GreenContext& operator=(GreenContext&& other) noexcept {
+    if (this != &other) {
+      // Clean up current resources
+      if (green_ctx_) {
+        CUcontext current = nullptr;
+        C10_CUDA_DRIVER_CHECK(
+            c10::cuda::DriverAPI::get()->cuCtxGetCurrent_(&current));
+        if (current == context_) {
+          TORCH_CHECK(
+              false,
+              "attempting to overwrite current green ctx "
+              "when it is active!");
+        }
+        C10_CUDA_DRIVER_CHECK(cuGreenCtxDestroy(green_ctx_));
+      }
+
+      // Take ownership of other's resources
+      green_ctx_ = std::exchange(other.green_ctx_, nullptr);
+      context_ = std::exchange(other.context_, nullptr);
+      parent_stream_ = std::exchange(other.parent_stream_, nullptr);
+    }
+    return *this;
+  }
+
+  ~GreenContext() noexcept {
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 12080
+    C10_CUDA_DRIVER_CHECK(
+        c10::cuda::DriverAPI::get()->cuGreenCtxDestroy_(green_ctx_));
+#else
+    TORCH_CHECK(false, "Green Context is only supported on CUDA 12.8+!");
+#endif
+  }
+
+  // Get the underlying CUDA context
+  CUcontext getContext() const {
+    return context_;
+  }
+
+  // Get the underlying green context
+  CUgreenCtx getGreenContext() const {
+    return green_ctx_;
+  }
+
+  // Make this context current
+  void makeCurrent() {
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 12080
+    auto current_stream = c10::cuda::getCurrentCUDAStream();
+    parent_stream_ = current_stream.stream();
+    cudaEvent_t ev;
+    C10_CUDA_CHECK(cudaEventCreate(&ev));
+    C10_CUDA_CHECK(cudaEventRecord(ev, current_stream));
+    CUcontext current = nullptr;
+    C10_CUDA_DRIVER_CHECK(
+        c10::cuda::DriverAPI::get()->cuCtxGetCurrent_(&current));
+    if (!current) {
+      C10_CUDA_DRIVER_CHECK(
+          c10::cuda::DriverAPI::get()->cuCtxSetCurrent_(context_));
+    } else {
+      C10_CUDA_DRIVER_CHECK(
+          c10::cuda::DriverAPI::get()->cuCtxPushCurrent_(context_));
+    }
+    // currently hardcodes the new green context to use the default stream
+    // TODO(eqy): consider creating a new stream if e.g., it allows interop
+    // with CUDA Graph captures etc.
+    C10_CUDA_CHECK(cudaStreamWaitEvent(NULL, ev, 0));
+    c10::cuda::setCurrentCUDAStream(c10::cuda::getDefaultCUDAStream());
+    C10_CUDA_CHECK(cudaEventDestroy(ev));
+#else
+    TORCH_CHECK(false, "Green Context is only supported on CUDA 12.8+!");
+#endif
+  }
+
+  void popCurrent() {
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 12080
+    // see above note about stream being hardcoded to the default stream
+    cudaEvent_t ev;
+    C10_CUDA_CHECK(cudaEventCreate(&ev));
+    C10_CUDA_CHECK(cudaEventRecord(ev, NULL));
```

Comment:
```suggestion
    CUDAEvent ev;
    ev.record(0);
```
Did you mean to hardcode NULL stream here, or should it be current_stream?